### ISSUE TITLE
LWT may commit incomplete updates which miss parts depending on the new schema if they were made before the schema fully propagated

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#12556]
Scenario:

1. Nodes {A, B, C}
2. Schema changes from v1 to v2, propagates to {A} 
3. LWT transaction starts on {A}, update depends on v2. Update is accepted on {A, C}
4. Node A restarts
5. LWT transaction starts on {B}, picks up previous update and drives it to completion. Downgrades its schema to v1 (after 15ebd5907177ba8709bc40ad2fa9ae773a9cf674), and commits it.

The downgrade of update to v1 can lose mutations which depend on v2, e.g. writes to new columns.

Before  15ebd5907177ba8709bc40ad2fa9ae773a9cf674 a similar problem could happen, but during learn phase. 

In general, downgrades from newer to older schema are not supported, so this could manifest in a different way. 

Solution proposal: https://github.com/scylladb/scylladb/issues/12556#issuecomment-1387161527

@gleb-cloudius @kostja

**Comment by gleb-cloudius** [comment]
@tgrabiec Any idea how to fix?

**Comment by avikivity** [comment]
We could record the schema in the promise, and then B can wait until it sees v2. Must use the schema history table for that. But the actual version may not ever be seen, since two concurrent updates may happen and B will only see their sum.

**Comment by avikivity** [comment]
Perhaps if we had linearizable schema changes it could be made to work.

Alternatively, A must make sure B has updated schema before it sends the mutation.

**Comment by gleb-cloudius** [comment]
> Perhaps if we had linearizable schema changes it could be made to work.

With linearizable schema changes we probably can. The problem is that even with it given two schema uuids we cannot know which one is newer and which is older since they are not time_uuids. If they would have been I see a simple (relatively) solution. 

> Alternatively, A must make sure B has updated schema before it sends the mutation.

How? :) The flow of the schema in Scylla is from a coordinator to replicas. But even if we could make replica reply with "upgrade your schema and retry" (which is the simple solution I thought about above) how do we know if the schema is newer or older?

**Comment by avikivity** [comment]
> > Perhaps if we had linearizable schema changes it could be made to work.
> 
> With linearizable schema changes we probably can. The problem is that even with it given two schema uuids we cannot know which one is newer and which is older since they are not time_uuids. If they would have been I see a simple (relatively) solution.

We have the schema history table, no?

> 
> > Alternatively, A must make sure B has updated schema before it sends the mutation.
> 
> How? :) The flow of the schema in Scylla is from a coordinator to replicas. But even if we could make replica reply with "upgrade your schema and retry" (which is the simple solution I thought about above) how do we know if the schema is newer or older?

Perhaps look for the newest timestamp in the mutation form?

It works with deletions since there will be a tombstone timestamp, but need to consider purged tombstones too. Maybe it's okay because we have deleted_columns?

**Comment by tgrabiec** [comment]
One possible solution would look like this.

paxos_state::prepare() response contains the original schema version of an update.

When LWT coordinator picks up updates which don't come from the current request, and their original schema is different than the current schema of the request, the coordinator synchronizes with one of those replicas. If it was behind, then it requests the updates again from paxos_state, this time with its latest schema, not the schema of the request. This schema is guaranteed to be at least as recent as the original schema of the update.

**Comment by gleb-cloudius** [comment]
> > > Perhaps if we had linearizable schema changes it could be made to work.
> > 
> > 
> > With linearizable schema changes we probably can. The problem is that even with it given two schema uuids we cannot know which one is newer and which is older since they are not time_uuids. If they would have been I see a simple (relatively) solution.
> 
> We have the schema history table, no?

We have column mapping table. Schema id is a clustering key. I do not see how we can figure out the order of schemas from this.

> 
> > > Alternatively, A must make sure B has updated schema before it sends the mutation.
> > 
> > 
> > How? :) The flow of the schema in Scylla is from a coordinator to replicas. But even if we could make replica reply with "upgrade your schema and retry" (which is the simple solution I thought about above) how do we know if the schema is newer or older?
> 
> Perhaps look for the newest timestamp in the mutation form?

We do not have one of the schemas already. We have only its id.

**Comment by kostja** [comment]
Fixed by:
commit 15ebd5907177ba8709bc40ad2fa9ae773a9cf674
Author: Gleb Natapov' via ScyllaDB development <scylladb-dev@googlegroups.com>
Date:   Thu Jan 12 12:35:42 2023 +0200

    lwt: upgrade stored mutations to the latest schema during prepare
    
    Currently they are upgraded during learn on a replica. The are two
    problems with this.  First the column mapping may not exist on a replica
    if it missed this particular schema (because it was down for instance)
    and the mapping history is not part of the schema. In this case "Failed
    to look up column mapping for schema version" will be thrown. Second lwt
    request coordinator may not have the schema for the mutation as well
    (because it was freed from the registry already) and when a replica
    tries to retrieve the schema from the coordinator the retrieval will fail
    causing the whole request to fail with "Schema version XXXX not found"
    
    Both of those problems can be fixed by upgrading stored mutations
    during prepare on a node it is stored at. To upgrade the mutation its
    column mapping is needed and it is guarantied that it will be present
    at the node the mutation is stored at since it is pre-request to store
    it that the corresponded schema is available. After that the mutation
    is processed using latest schema that will be available on all nodes.
    
    Fixes #10770
    
    Message-Id: <Y7/ifraPJghCWTsq@scylladb.com>

**Comment by tgrabiec** [comment]
This issue was created after https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674, and mentions it in the commit message, so we were aware of it. Why do you think it fixes this issue?

**Comment by kostja** [comment]
I believe https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674 never downgrades the schema. It will automatically retrieve the latest schema if it is available at any peer during paxos repair, and upgrade the proposer. But it will not downgrade the mutation if it fails to find the relevant schema, the code throws on_internal_error in this case. If it can manage to commit the Paxos repair on a majority because the majority still has the old schema, I don't think it's an issue we should track as a lot update. The quorum rules are preserved.

What you perhaps could wish for is a raft read  barrier *whenever* we discover an unknown schema version. For that, we need a better schema version. E.g., with Raft based schema changes, the situation you describe is impossible in a cluster of A, B, C. It would take a cluster of at least 5 nodes to make it possible to commit a schema change without affecting B and C. To fix it even in a cluster of A, B, C, D, E, B, when it begins the repair, and observes a later schema version, could first perform a raft read barrier to get it. 

Another matter is the bug title. The scenario you discuss is an uncertainty scenario. Obviously the schema has not propagated to all of the nodes at the time of the updates, and we haven't confirmed the update to the client. So why do you believe we *have to* preserve the schema version of A? From the observer point of view, it was never promised that  A's schema version would be used. If there was at least one successful update on this replica set which would use A's version, this version would propagate to the majority of replicas.

See, here's my problem with hypothetical bug scenarios. We could speculate how bad they are for a very long time. But we would never be able to know for sure if there is no bug in our speculation. We would not be able to assess their impact. So I accept I didn't notice you reported it after the original issue, as a case which it hypothetically doesn't address. But my reasons  why we should not be tracking this issue are above. 

Instead of having to track this complex scenario, I would open an issue which would demand linearization of schema changes and LWT (or strongly consistent tablet writes). As to the case above, I think we never promised it to anyone, so should not track it as a bug.

**Comment by tgrabiec** [comment]
> I believe [15ebd59](https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674) never downgrades the schema. 

It can downgrade the schema in paxos_state::prepare():

https://github.com/scylladb/scylladb/blob/c65bd01174c1a1089a056c56725c5ffc6f7be2aa/service/paxos/paxos_state.cc#L116-L119

The `schema` comes from `_cmd` on the coordinator side, so it's chosen when the LWT request starts. It's the local schema of the coordinator. It may be older than the schema of p->update. The coordinator doesn't know that it has an older schema, and will not know, because replicas respond with mutations already downgraded. 

> It will automatically retrieve the latest schema if it is available at any peer during paxos repair, and upgrade the proposer. But it will not downgrade the mutation if it fails to find the relevant schema, the code throws on_internal_error in this case.

The repair is performed using most_recent_proposal which comes from responses received from replicas during prepare(), which are already downgraded (in paxos_state::prepare()) to the schema used by the coordinator. This will be the same schema used during learn, so the following check can't fail, we'll commit the downgraded mutation: 

https://github.com/scylladb/scylladb/blob/c65bd01174c1a1089a056c56725c5ffc6f7be2aa/service/paxos/paxos_state.cc#L217-L219


> 
> What you perhaps could wish for is a raft read barrier _whenever_ we discover an unknown schema version. For that, we need a schema version which is monotonic and ideally includes the timesetamp.

There's also a non-raft solution proposed in comments here.

> 
> Another observation is about a potential impact, it's very low, I think it's nearly impossible. With Raft based schema changes, tIt would take a cluster of at least 5 nodes to make it possible to commit a schema change without affecting B and C. 

Raft doesn't make it less likely because schema is still applied at different times on different nodes, so can race with LWT requests. The scenario only requires node A to go down in order to interrupt the LWT flow, not to prevent schema propagation.

> 
> Another matter is the bug title. The scenario you discuss is an uncertainty scenario. Obviously the schema has not propagated to all of the nodes at the time of the updates, and we haven't confirmed the update to the client. So why do you believe we _have to_ preserve the schema version of A? From the observer point of view, it was never promised that A's schema version would be used. 

The bug is not about guaranteeing that A's schema version is used, but about partial success. It's not guaranteed which schema version is used, but the update should stick to one schema and either fail or be applied fully, never partially. 

> 
> See, here's my problem with hypothetical bug scenarios. We could speculate how bad they are for a very long time. But we would never be able to know for sure if there is no bug in our speculation. We would not be able to assess their impact. So I accept I didn't notice you reported it after the original issue, as a case which it hypothetically doesn't address. But my reasons why we should not be tracking this issue are above.
> 
> Instead of having to track this complex scenario, I would open an issue which would demand linearization of schema changes and LWT (or strongly consistent tablet writes).

I don't know what you mean by "linearization of schema changes and LWT", but we can consider it as an alternative fix if there is an issue for it.

**Comment by kostja** [comment]
> > I believe [15ebd59](https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674) never downgrades the schema.
> 
> It can downgrade the schema in paxos_state::prepare():
> 
> https://github.com/scylladb/scylladb/blob/c65bd01174c1a1089a056c56725c5ffc6f7be2aa/service/paxos/paxos_state.cc#L116-L119
> 
> The `schema` comes from `_cmd` on the coordinator side, so it's chosen when the LWT request starts. It's the local schema of the coordinator. It may be older than the schema of p->update. The coordinator doesn't know that it has an older schema, and will not know, because replicas respond with mutations already downgraded.
> 

First, I never thought "upgrade_if_needed()" could actually **downgrade** the schema. Thanks for pointing out. One more reason we need to have schema version monotonicity. 
> > Another matter is the bug title. The scenario you discuss is an uncertainty scenario. Obviously the schema has not propagated to all of the nodes at the time of the updates, and we haven't confirmed the update to the client. So why do you believe we _have to_ preserve the schema version of A? From the observer point of view, it was never promised that A's schema version would be used.
> 
> The bug is not about guaranteeing that A's schema version is used, but about partial success. It's not guaranteed which schema version is used, but the update should stick to one schema and either fail or be applied fully, never partially.

Do you mean that different nodes can apply the mutation at different schema versions? But how, if it is downgraded first during repair? The apply didn't happen, the round will be re-prepared, so we are going to use the old version everywhere. 

> > See, here's my problem with hypothetical bug scenarios. We could speculate how bad they are for a very long time. But we would never be able to know for sure if there is no bug in our speculation. We would not be able to assess their impact. So I accept I didn't notice you reported it after the original issue, as a case which it hypothetically doesn't address. But my reasons why we should not be tracking this issue are above.
> > Instead of having to track this complex scenario, I would open an issue which would demand linearization of schema changes and LWT (or strongly consistent tablet writes).
> 
> I don't know what you mean by "linearization of schema changes and LWT", but we can consider it as an alternative fix if there is an issue for it.

My point is that I don't even want to explore the infinite dark corners of how lack of schema version monotonicity and schema version fencing can go wrong. I made a proposal in the https://docs.google.com/document/d/13BTHL42vZBOuIwustZ_GFj_lfVcp77MPwm2it3uMGu0/edit and I think we're better of focusing on a systematic solution rather than explore how we could fix it "without Raft". Especially in absence of a test case. Essentially exploring in this direction has no observable benefit.

**Comment by kostja** [comment]
@tgrabiec thinking more of it, I believe you misunderstand how a lookup in schema version history works. We're trying to look up the mutation vresion in the history table, and if we succeed, we know it's an outdated version. But if it's a new version, and our history table doesn't contain it, the lookup will fail and we won't upgrade. We'll fail on the internal error branch.

**Comment by tgrabiec** [comment]
On Mon, Apr 10, 2023 at 10:53 AM Konstantin Osipov ***@***.***>
wrote:

> I believe 15ebd59
> <https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674>
> never downgrades the schema.
>
> It can downgrade the schema in paxos_state::prepare():
>
>
> https://github.com/scylladb/scylladb/blob/c65bd01174c1a1089a056c56725c5ffc6f7be2aa/service/paxos/paxos_state.cc#L116-L119
>
> The schema comes from _cmd on the coordinator side, so it's chosen when
> the LWT request starts. It's the local schema of the coordinator. It may be
> older than the schema of p->update. The coordinator doesn't know that it
> has an older schema, and will not know, because replicas respond with
> mutations already downgraded.
>
> First, I never thought "upgrade_if_needed()" could actually *downgrade*
> the schema. Thanks for pointing out. One more reason we need to have schema
> version monotonicity.
>
It's not meant to be used to downgrade, and in some cases may throw, but
it's not guaranteed that it will fail.

> Another matter is the bug title. The scenario you discuss is an
> uncertainty scenario. Obviously the schema has not propagated to all of the
> nodes at the time of the updates, and we haven't confirmed the update to
> the client. So why do you believe we *have to* preserve the schema
> version of A? From the observer point of view, it was never promised that
> A's schema version would be used.
>
> The bug is not about guaranteeing that A's schema version is used, but
> about partial success. It's not guaranteed which schema version is used,
> but the update should stick to one schema and either fail or be applied
> fully, never partially.
>
> Do you mean that different nodes can apply the mutation at different
> schema versions? But how, if it is downgraded first during repair? The
> apply didn't happen, the round will be re-prepared, so we are going to use
> the old version everywhere.
>
No, I don't mean that replicas will use different schemas, but that request
coordinator will use a different schema than the replication layer. The
coordinator uses the new schema to build the mutations, that's why the
statement passes validation, but replication will use the old schema,
that's why downgrade happens. If the coordinator used the same schema as
replication then the request would either fail during mutation building or
would be replicated fully.


> See, here's my problem with hypothetical bug scenarios. We could speculate
> how bad they are for a very long time. But we would never be able to know
> for sure if there is no bug in our speculation. We would not be able to
> assess their impact. So I accept I didn't notice you reported it after the
> original issue, as a case which it hypothetically doesn't address. But my
> reasons why we should not be tracking this issue are above.
> Instead of having to track this complex scenario, I would open an issue
> which would demand linearization of schema changes and LWT (or strongly
> consistent tablet writes).
>
> I don't know what you mean by "linearization of schema changes and LWT",
> but we can consider it as an alternative fix if there is an issue for it.
>
> My point is that I don't even want to explore the infinite dark corners of
> how lack of schema version monotonicity and schema version fencing can go
> wrong. I made a proposal in the
> https://docs.google.com/document/d/13BTHL42vZBOuIwustZ_GFj_lfVcp77MPwm2it3uMGu0/edit
> and I think we're better of focusing on a systematic solution rather than
> explore how we could fix it "without Raft". Especially in absence of a test
> case. Essentially exploring in this direction has no observable benefit.
>
I agree that if we can rely on schema version monotonicity, the fix is
pretty simple. We don't need to implement the full spec to fix this bug,
just pull on the replica side.

**Comment by tgrabiec** [comment]
Which lookup are you referring to? If it's about the lookup on replica-side
in handle_paxos_prepare(), then it uses get_schema_for_read() which should
succeed for any schema version, pulling the schema from the coordinator if
necessary.

**Comment by DoronArazii** [comment]
Missed the train to 5.2, pushing to 5.3

**Comment by kostja** [comment]
Fix: now that we have schema version monotonicity, when performing a repair we should check if the current node has an older schema than the version used in the mutation sent by repair and fetch the latest raft schema in this case.

Generally it's nearly impossible this would happen (without heavy sleep injections).

**Comment by kbr-scylla** [comment]
> now that we have schema version monotonicity

We don't. Not until we get rid of gossip-based RECOVERY, enforcing Raft in cluster 100% of the time.

**Comment by kbr-scylla** [comment]
Marking as `blocked` by https://github.com/scylladb/scylladb/issues/15422

**Comment by kbr-scylla** [comment]
https://github.com/scylladb/scylladb/issues/15422 is not enough though, I opened https://github.com/scylladb/scylladb/issues/19176

**Comment by tgrabiec** [comment]
\cc @gusev-p
